cfd_protocol/src/lib.rs

```diff
@@ -162,13 +520,375 @@ mod tests {
+    ) -> Result<Transaction> {
+        let satisfier = {
+            let mut satisfier = HashMap::with_capacity(2);
+
+            // The order in which these are inserted doesn't matter
+            satisfier.insert(maker_pk, (maker_sig, SigHashType::All));
+            satisfier.insert(taker_pk, (taker_sig, SigHashType::All));
+
+            satisfier
+        };
+
+        let mut tx_refund = self.inner;
+        self.commit_output_descriptor
+            .satisfy(&mut tx_refund.input[0], satisfier)?;
+
+        Ok(tx_refund)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct CommitTransaction {
+    inner: Transaction,
+    descriptor: Descriptor<PublicKey>,
+    amount: Amount,
+    sighash: SigHash,
+    lock_descriptor: Descriptor<PublicKey>,
+}
+
+impl CommitTransaction {
+    fn new(
+        lock_tx: &LockTransaction,
+        (maker_own_pk, maker_rev_pk, maker_publish_pk): (PublicKey, PublicKey, PublicKey),
+        (taker_own_pk, taker_rev_pk, taker_publish_pk): (PublicKey, PublicKey, PublicKey),
+    ) -> Result<Self> {
+        // FIXME: Fee to be paid by leftover lock output
+        let amount = lock_tx.amount();
+
+        let lock_input = TxIn {
+            previous_output: lock_tx.lock_outpoint(),
+            ..Default::default()
+        };
+
+        let descriptor = Self::build_descriptor(
+            (maker_own_pk, maker_rev_pk, maker_publish_pk),
+            (taker_own_pk, taker_rev_pk, taker_publish_pk),
+        );
+
+        let output = TxOut {
+            value: lock_tx.amount().as_sat(),
+            script_pubkey: descriptor
+                .address(Network::Regtest)
+                .expect("can derive address from descriptor")
+                .script_pubkey(),
+        };
+
+        let inner = Transaction {
+            version: 2,
+            lock_time: 0,
+            input: vec![lock_input],
+            output: vec![output],
+        };
+
+        let sighash = SigHashCache::new(&inner).signature_hash(
+            0,
+            &lock_tx.descriptor().script_code(),
+            lock_tx.amount().as_sat(),
+            SigHashType::All,
+        );
+
+        Ok(Self {
+            inner,
+            descriptor,
+            lock_descriptor: lock_tx.descriptor(),
+            amount,
+            sighash,
+        })
+    }
+
+    fn encsign(&self, sk: SecretKey, publish_them_pk: &PublicKey) -> Result<EcdsaAdaptorSignature> {
+        Ok(EcdsaAdaptorSignature::encrypt(
+            SECP256K1,
+            &secp256k1_zkp::Message::from_slice(&self.sighash).expect("sighash is valid message"),
+            &sk,
+            &publish_them_pk.key,
+        ))
+    }
+
+    fn encverify(
+        &self,
+        encsig: &EcdsaAdaptorSignature,
+        signing_pk: &PublicKey,
+        encryption_pk: &PublicKey,
+    ) -> Result<()> {
+        encsig.verify(
+            SECP256K1,
+            &secp256k1_zkp::Message::from_slice(&self.sighash).expect("sighash is valid message"),
```

Comment:
I think all the hash types of the `rust-bitcoin` ecosystem have fail-safe `From` implementations via this `ThirtyTwoByteHash` marker trait?

https://docs.rs/secp256k1/0.20.2/src/secp256k1/lib.rs.html#504-509